### PR TITLE
Add checkup for toml keys

### LINKED ISSUE #401: Providing warnings/error message when TOML file has unknown keys/values
For now it just doesn't do anything and it's a bit complicated to guess what's wrong sometimes.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #401

### CODE REVIEW

**Comment by sdroege** [comment]
:+1:

**Comment by EPashkin** [comment]
Good idea, but IMHO better `error!` instead panic.

**Comment by GuillaumeGomez** [comment]
`panic!` seemed to be the way to do it but I'll replace with `error!` if you prefer this last one. :)

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
@GuillaumeGomez You will don't plan to add this for other places: functions, parameters etc.?

**Comment by GuillaumeGomez** [comment]
I can as well. I'll do it this evening.

**Comment by GuillaumeGomez** [comment]
@EPashkin: I just need a confirmation: you want me to add the same check for each line in here:

```rust
let functions = Functions::parse(toml_object.lookup("function"), &name);
let signals = Signals::parse(toml_object.lookup("signal"), &name);
let members = Members::parse(toml_object.lookup("member"), &name);
let properties = Properties::parse(toml_object.lookup("property"), &name);
```

?

**Comment by EPashkin** [comment]
To make PR complete need add `check_unwanted` to all `parse` functions, or it will check only small subset.

**Comment by GuillaumeGomez** [comment]
Ok, let's do it then!

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/config/child_properties.rs [hunk 1]
```diff
@@ -12,6 +12,9 @@ pub struct ChildProperty {
 
 impl Parse for ChildProperty {
     fn parse(toml: &Value, object_name: &str) -> Option<ChildProperty> {
+        toml.check_unwanted(&["name", "type", "doc_hidden"],
+                            &format!("child property {}", object_name));
```
Comment (EPashkin) [inline src/config/child_properties.rs]: Please, move this after we checked name.

Comment (GuillaumeGomez) [inline src/config/child_properties.rs]: Yes sir!

**Comment by GuillaumeGomez** [comment]
Done.

File: src/config/functions.rs [hunk 2]
```diff
@@ -28,6 +28,9 @@ impl Parse for Parameter {
                 return None;
             }
         };
+        toml.check_unwanted(&["const", "nullable", "length_of", "name"],
```
Comment (EPashkin) [inline src/config/functions.rs]: "pattern" also needed for this and other with `Ident::parse`.
Strange that it don't failed on gtk.

Comment (EPashkin) [inline src/config/functions.rs]: Hm, "pattern" usable for parameters?

**Comment by EPashkin** [comment]
For gdk:
```
"member Gdk.EventType": Unknown key: ["name"]
"ident Gdk.EventType": Unknown key: ["alias"]
"child properties Gdk.EventType": Unknown key: ["member"]
```
Glib:
```
ERROR:gir::config::error: "ident GLib.*": Unknown key: ["cfg_condition"]
ERROR:gir::config::error: "ident GLib.*": Unknown key: ["cfg_condition"]
ERROR:gir::config::error: "member GLib.ChecksumType": Unknown key: ["name"]
ERROR:gir::config::error: "child properties GLib.ChecksumType": Unknown key: ["member"]
ERROR:gir::config::error: "child properties GLib.MainContext": Unknown key: ["concurrency"]
ERROR:gir::config::error: "child properties GLib.MainLoop": Unknown key: ["concurrency"]
ERROR:gir::config::error: "child properties GLib.Source": Unknown key: ["concurrency"]
```

File: src/config/functions.rs (hunk 2 above)
Comment (GuillaumeGomez) [inline src/config/functions.rs]: > Hm, "pattern" usable for parameters?

Apparently...

**Comment by GuillaumeGomez** [comment]
Ah, I need to update my gir version I think. ;)

File: src/config/functions.rs (hunk 2 above)
Comment (EPashkin) [inline src/config/functions.rs]: It used in Pango

**Comment by GuillaumeGomez** [comment]
Updated.

**Comment by EPashkin** [comment]
Still gdk: `ERROR:gir::config::error: "child properties Gdk.EventType": Unknown key: ["member"]`
and glib
```
ERROR:gir::config::error: "child properties GLib.ChecksumType": Unknown key: ["member"]
ERROR:gir::config::error: "child properties GLib.MainContext": Unknown key: ["concurrency"]
ERROR:gir::config::error: "child properties GLib.MainLoop": Unknown key: ["concurrency"]
ERROR:gir::config::error: "child properties GLib.Source": Unknown key: ["concurrency"]
```

**Comment by GuillaumeGomez** [comment]
Sorry, I failed my run I suppose. Retrying.

**Comment by GuillaumeGomez** [comment]
*Should* be good this time.

**Comment by EPashkin** [comment]
`Ident::parse` don't need check

**Comment by GuillaumeGomez** [comment]
You want me to remove the whole `check_unwanted` from `Ident::parse` or you're talking about a specific key?

**Comment by EPashkin** [comment]
Whole `check_unwanted` or error shows 2 times:
```
ERROR:gir::config::error: "ident Gtk.Widget": Unknown key: ["ttt"]
ERROR:gir::config::error: "function Gtk.Widget": Unknown key: ["ttt"]
```

**Comment by GuillaumeGomez** [comment]
Done.

**Comment by EPashkin** [comment]
Now it works fine on gtk-rs.
Want squash or prefer merge as is?

**Comment by GuillaumeGomez** [comment]
You tell me. You're the maintainer so as you prefer.

File: src/config/child_properties.rs [hunk 3]
```diff
@@ -56,6 +59,11 @@ pub struct ChildProperties {
 
 impl Parse for ChildProperties {
     fn parse(toml_object: &Value, object_name: &str) -> Option<ChildProperties> {
+        toml_object.check_unwanted(&["child_name", "child_type", "child_prop", "function", "name",
+                                     "status", "signal", "property", "version", "module_name",
+                                     "member", "concurrency"],
```
Comment (EPashkin) [inline src/config/child_properties.rs]: Seems this check also don't need as it checked in `parse_object`.
maybe `parse_object` need comment that it check `child_*` for this?

**Comment by GuillaumeGomez** [comment]
Done!

File: src/config/members.rs [hunk 4]
```diff
@@ -15,6 +15,9 @@ pub struct Member {
 
 impl Parse for Member {
     fn parse(toml: &Value, object_name: &str) -> Option<Member> {
+        toml.check_unwanted(&["member", "alias", "version", "name", "pattern"],
```
Comment (EPashkin) [inline src/config/members.rs]: "member" don't needed

File: src/config/properties.rs [hunk 5]
```diff
@@ -16,6 +16,9 @@ pub struct Property {
 
 impl Parse for Property {
     fn parse(toml: &Value, object_name: &str) -> Option<Property> {
+        toml.check_unwanted(&["property", "ignore", "version", "name", "pattern"],
```
Comment (EPashkin) [inline src/config/properties.rs]: "property" also unneeded

File: src/config/members.rs (hunk 4 above)
Comment (EPashkin) [inline src/config/members.rs]: and better move after Ident parse

File: src/config/properties.rs (hunk 5 above)
Comment (EPashkin) [inline src/config/properties.rs]: also better move down

**Comment by EPashkin** [comment]
All checked, squash, please.

**Comment by GuillaumeGomez** [comment]
Updated and squashed.

**Comment by EPashkin** [comment]
Thanks, @GuillaumeGomez

**Pull request merged**